go/datas/pull.go

```diff
@@ -29,259 +28,63 @@ const bytesWrittenSampleRate = .10
-					// Hook in here to estimate the bytes written to disk during pull (since
-					// srcChan contains all chunks to be written to the sink). Rather than measuring
-					// the serialized, compressed bytes of each chunk, we take a 10% sample.
-					// There's no immediately observable performance benefit to sampling here, but there's
-					// also no appreciable loss in accuracy, so we'll keep it around.
-					takeSample := rand.Float64() < bytesWrittenSampleRate
-					srcResChan <- traverseSource(srcRef, srcDB, sinkDB, takeSample)
-				case sinkRef := <-sinkChan:
-					sinkResChan <- traverseSink(sinkRef, srcDB)
-				case comRef := <-comChan:
-					comResChan <- traverseCommon(comRef, sinkHeadRef, srcDB)
-				case <-done:
-					workerWg.Done()
-					return
-				}
-			}
-		}()
-	}
-	for i := 0; i < concurrency; i++ {
-		traverseWorker()
+		return // already up to date
 	}
 
 	var doneCount, knownCount, approxBytesWritten uint64
-	updateProgress := func(moreDone, moreKnown, moreBytesRead, moreApproxBytesWritten uint64) {
+	updateProgress := func(moreDone, moreKnown, moreApproxBytesWritten uint64) {
 		if progressCh == nil {
 			return
 		}
 		doneCount, knownCount, approxBytesWritten = doneCount+moreDone, knownCount+moreKnown, approxBytesWritten+moreApproxBytesWritten
-		progressCh <- PullProgress{doneCount, knownCount + uint64(srcQ.Len()), approxBytesWritten}
+		progressCh <- PullProgress{doneCount, knownCount, approxBytesWritten}
 	}
-
-	sampleSize := uint64(0)
-	sampleCount := uint64(0)
-	for !srcQ.Empty() {
-		srcRefs, sinkRefs, comRefs := planWork(srcQ, sinkQ)
-		srcWork, sinkWork, comWork := len(srcRefs), len(sinkRefs), len(comRefs)
-		if srcWork+comWork > 0 {
-			updateProgress(0, uint64(srcWork+comWork), 0, 0)
-		}
-
-		// These goroutines send work to traverseWorkers, blocking when all are busy. They self-terminate when they've sent all they have.
-		go sendWork(srcChan, srcRefs)
-		go sendWork(sinkChan, sinkRefs)
-		go sendWork(comChan, comRefs)
-		//  Don't use srcRefs, sinkRefs, or comRefs after this point. The goroutines above own them.
-
-		for srcWork+sinkWork+comWork > 0 {
-			select {
-			case res := <-srcResChan:
-				for _, reachable := range res.reachables {
-					srcQ.PushBack(reachable)
-				}
-				if res.writeBytes > 0 {
-					sampleSize += uint64(res.writeBytes)
-					sampleCount += 1
-				}
-				srcWork--
-
-				updateProgress(1, 0, uint64(res.readBytes), sampleSize/uint64(math.Max(1, float64(sampleCount))))
-			case res := <-sinkResChan:
-				for _, reachable := range res.reachables {
-					sinkQ.PushBack(reachable)
-				}
-				sinkWork--
-			case res := <-comResChan:
-				isHeadOfSink := res.readHash == sinkHeadRef.TargetHash()
-				for _, reachable := range res.reachables {
-					sinkQ.PushBack(reachable)
-					if !isHeadOfSink {
-						srcQ.PushBack(reachable)
-					}
-				}
-				comWork--
-				updateProgress(1, 0, uint64(res.readBytes), 0)
+	var sampleSize, sampleCount uint64
+
+	cc := newCompletenessChecker()
```

Comment:
i think we're in danger of having two completeness checkers operating. if the chunkStore is an httpChunkStore, then it will already have been wrapped in a validatingChunkStore